gz-6: Added basic incorporation of bootstrap and redesigned the layout of the homepage

NickMcL opened the pull request:
This ticket incorporates bootstrap into the page templates to improve the interface design of the application. In addition, the overall layout of the homepage was changed to improve the interface, and some adjustments were made to the code to eliminate parts that are no longer used and to adjust the handling of bad user data.

Although bootstrap is now successfully incorporated on all of the templates, most of the enhancements from it are basic changes that avoid using some of bootstrap's features to their full extent. This still allowed for a dramatic improvement in the design of the app, but some areas could still use more work on improving the design. Most notably, the form pages could definitely use more work, but since it seems that it will require additional django applications to allow the forms to fully utilize bootstrap, I will probably work on those pages later after most of the app's functions are working.

Since the design of the app is now at least at an acceptable level, I will probably work on optimizing and completing the app's functions before spending any more time working to improve the visual design of the app.

NickMcL committed:
Set up settings file for local testing

NickMcL committed:
Eliminated unnecessary CSS

NickMcL committed:
Split the No Association list rendered to the home page into to lists of equal length in order to display the information in two columns of entries

NickMcL committed:
Revert "Split the No Association list rendered to the home page into to lists of equal length in order to display the information in two columns of entries"

This reverts commit 0cd34a4045abf1fa79c8fd8f4f7c6bf2909bd662.

NickMcL committed:
No associations table now has two columns of entries

NickMcL committed:
Bootstrap testing: Enabled bootstrap to be linked to the templates by adjusting the statics settings in django

NickMcL committed:
gz-6: Merged the branch that I was using to test out configuring bootstrap with django. This ticket is now going to be for doing basic incorporation of bootstrap into the app.

NickMcL committed:
gz-6: Started aligning the table html to the specifications in bootstrap

NickMcL committed:
gz-6: Reformatted the three main association tables using bootstrap

NickMcL committed:
gz-6: Fixed the Quick Reference tables with bootstrap so that they display on the home page properly. Also, fixed the centering of all of the tables.

NickMcL committed:
gz-6: Removed template files that are no longer used in the application.

NickMcL committed:
gz-6: Added basic bootstrap styling to the login form.

NickMcL committed:
gz-6: Added basic bootstrap styling to the change form.

NickMcL committed:
gz-6: Added basic bootstrap styling to the confirmation and error page templates

NickMcL committed:
gz-6: Removed the code that gathered the requester information for Zendesk and GitHub tickets because this information is no longer used after redesigning the home page with bootstrap.

NickMcL committed:
gz-6: When testing out the new design with bootstrap, it became apparent that the error handling was not working properly in the code, so I fixed the code to allow for bad user information to still pass through the code and have the appropriate messages rendered to the home page in the new design.

NickMcL committed:
Settings adjusted for deployment to heroku instead of local testing

winhamwr commented on associations/templates/associations/change.html:
```diff
@@ -1,21 +1,26 @@
 <html>
 <head>
+<link rel="stylesheet" href="{{ STATIC_URL }}css/bootstrap.css">
```
Instead of including this everywhere, you're better of using a base template that includes bootstrap and extending that template.

winhamwr replied:
Also, it looks like you have the minified and -responsive versions of bootstrap included. Are they used anywhere?

NickMcL replied:
No, I'm just using the normal bootstrap.css, so I can delete those extra files.

NickMcL replied:
I deleted the unused CSS files and created a base template ("base.html") that includes bootstrap.

winhamwr commented on associations/templates/associations/change.html:
```diff
@@ -1,21 +1,26 @@
 <br/>
 
-<form action="/change/" method="post">{% csrf_token %}
+<form class="well" action="/change/" method="post">{% csrf_token %}
```
Don't hard-code this URL. You can leave the action blank if it should always hit the same page, or you should use a url template tag.

NickMcL replied:
Fixed. I deleted the action parameter on the forms.

winhamwr commented on associations/templates/associations/confirm.html:
```diff
@@ -1,23 +1,36 @@
 <head>
+<link rel="stylesheet" href="{{ STATIC_URL }}css/bootstrap.css">
+
+<style type="text/css">
```
Pull this and the rest of your CSS in to a separate stylesheet.

NickMcL replied:
Fixed. I moved the CSS into the stylesheet "base.css".

winhamwr commented on associations/templates/associations/home.html:
```diff
@@ -141,34 +74,110 @@
+</div>
+
+<div class="span6">
+    <h3>Half Open Associations</h3>
```
Need to work on the wording here and on the app name in general. "Associations" is a classic horrible name for an app or variable. Completely ambiguous. You'll want something more specific and descriptive (eg. "Open Enhancement Requests", "enhancement_tracking", "Enhancements Requiring Notification", etc). The labels should be from the user's perspective and talk about what they actually want to use this tool for (tracking product enhancement against customer requests and then notifying customers upon completion).

NickMcL replied:
To try to address this issue, I renamed the "associations" app to "enhancement_tracking" in addition to renaming several functions and variables with the app.

winhamwr commented on associations/templates/associations/login.html:
```diff
@@ -1,17 +1,24 @@
+
 <h2>Existing User</h2>
-<form action="/" method="post">{% csrf_token %}
+<form class="well" action="/" method="post">{% csrf_token %}
```
Hard-coded URL

NickMcL replied:
Fixed. I deleted the action parameter on the forms.

winhamwr commented on associations/templates/associations/nope.html:
```diff
@@ -12,24 +22,28 @@
-                <br/><a href="/main/">Return to Homepage</a>
+				<br/>
+				<a href="/home/">Return to Homepage</a>
             {% else %}
```
Hard-coded URL

NickMcL replied:
Fixed. I replaced all of the hard-coded URLs with the url template tag.

winhamwr commented on associations/templates/associations/nope.html:
```diff
@@ -1,10 +1,20 @@
 <html>
```
"nope" is not a good name for a login page, and you should be using the login page to handle login errors. Instead of doing customer `nope_num` error handling, use Django's form system and error handling instead. There's a built in contrib.auth form that handles this, or you can use one of several open source django apps that help here. No need to re-invent this wheel.

NickMcL replied:
Yeah, I had a feeling there was a much better way to do error handling than my makeshift solution. I'll do some research into those other methods.

winhamwr commented on associations/views.py:
```diff
@@ -254,6 +252,9 @@ def api_calls(request):
             r_op = requests.get(url,
                                 params={'state': 'open', 'since': limit_str},
                                 auth=(user.git_name, user.git_pass))
+            if type(r_op.json) == type({}):
```
This is a strange error check. Why are we check this versus a dictionary type? Generally in a dynamically typed language, if you're doing a type comparison, you're doing it wrong. Can't you just do:

```
if not r_op.json:
```

NickMcL replied:
In GitHub's API, when you request tickets from a repository, it returns a list of dictionaries that holds the information for each individual ticket, but if there was an error in the request (such as "Bad authentication" or similar), GitHub instead returns a dictionary with a key "message" that holds the error message. Since the only time GitHub ever returns a dictionary instead of a list indicates that there was an error in the request, I thought checking the type of the returned object to see if it was a dictionary would be a decent method for checking if GitHub had an error in obtaining the tickets requested.

`if not r_op.json:` would not work because `r_op.json` would still contain something (the error message) if GitHub had an error in obtaining the tickets.

There might be a better way to do this, but checking the types was just the first solution I came up with.

winhamwr replied:
Are you sure they they don't provide a different HTTP status code for requests that are unsuccessful? You would be better off checking that you got a non-200 status code for your error case, as that's the way HTTP works. Also, in the hypothetical case that they didn't (pretty sure they do), I would expect several lines of comments explaining strange code like that to handle the case.

NickMcL replied:
As expected, they do use different HTTP status codes for requests that are unsuccessful, so I replaced the data type checking with code that utilizes those status codes.

winhamwr commented on associations/views.py:
```diff
@@ -294,6 +298,9 @@ def api_calls(request):
         url = '%s/api/v2/tickets.json?page=%s' % (user.zen_url, page)
```
Still need to pull this out in to a constant.

winhamwr replied:
This would also be a good place for a function called `_get_zendesk_tickets` that encapsulates this logic.

NickMcL replied:
I plan to rewrite this section of code in a future ticket because it does not function the way I want it to, so I made a TODO comment there for now.

winhamwr commented on settings.py:
```diff
@@ -72,6 +72,7 @@
     # Put strings here, like "/home/html/static" or "C:/www/django/static".
     # Always use forward slashes, even on Windows.
     # Don't forget to use absolute paths, not relative paths.
+    "/home/nick/GitZen/statics",
```
Don't hard-code this or the project will only work on your computer. Use something like:

```
import os.path

os.path.join(os.path.abspath(path.dirname(__file__)), 'statics')
```

NickMcL replied:
The settings file actually already has a constant defined for the root path called `PROJECT_ROOT`, so I could probably just use that to fix this.

NickMcL replied:
Fixed.

NickMcL committed:
Fixed hard-coded file path for statics in settings

NickMcL committed:
gz-6: Removed unused bootstrap files

NickMcL committed:
Removed unused bootstrap files

NickMcL committed:
gz-6: Moved CSS into a seperate stylesheet

NickMcL committed:
gz-6: Created a base html template that is extended by all of the teplates and includes the external CSS and base page structure.

NickMcL committed:
Merge branch 'gz-6' of github.com:FriedRice/GitZen into gz-6

NickMcL committed:
gz-6: Removed the unnecessary action parameter from the login and change forms.

NickMcL committed:
gz-6: Added RequestContext to the confirmation and error render functions so that the templates work with the CSS links in the base template.

NickMcL committed:
gz-6: Named all of the urls and then used the url template tag and reverse to remove all hard-coded urls in both the templates and views file.

NickMcL commented:
Since it was brought up in the last pull request, I eliminated all of the hard-coded URLs in the views.py file by using the `reverse` function to retrieve URLs instead.

NickMcL committed:
gz-6: Renamed the Association app and several functions and variables within it to focus around the more user relatable theme of 'Enhancement tracking' rather than the ambiguous previous theme of 'Associations'

NickMcL committed:
gz-6: Changed the check for errors in the GitHub API request to use status codes instead of type checks

NickMcL commented:
I think I've addressed all of your comments except for the form authentication and Zendesk API requests issue, and I plan to address both of those in separate tickets in the future since they are larger changes.

winhamwr commented:
I created #10 to track the django login form thing. Tomorrow, we should talk about using [hub](https://github.com/defunkt/hub) to turn your issues in to pull requests as you finish them. That way there's not a separate issue and pull request.

Nice work on these changes. Things are really moving along. I'm hopeful that next week you can start doing some training and get people up on the system.
